kernel/src/loader/mod.rs

```diff
@@ -144,6 +148,87 @@ where
     Ok(GuestAddress(ehdr.e_entry as usize))
 }
 
+#[cfg(target_arch = "aarch64")]
+pub fn load_kernel<F>(
+    guest_mem: &GuestMemory,
+    kernel_image: &mut F,
+    start_address: usize,
+) -> Result<GuestAddress>
+where
+    F: Read + Seek,
+{
+    /* Kernel boot protocol is specified in the kernel docs
+    Documentation/arm/Booting and Documentation/arm64/booting.txt.
+
+    ======aarch64 kernel header========
+    u32 code0;			/* Executable code */
+    u32 code1;			/* Executable code */
+    u64 text_offset;		/* Image load offset, little endian */
+    u64 image_size;		/* Effective Image size, little endian */
+    u64 flags;			/* kernel flags, little endian */
+    u64 res2	= 0;		/* reserved */
+    u64 res3	= 0;		/* reserved */
+    u64 res4	= 0;		/* reserved */
+    u32 magic	= 0x644d5241;	/* Magic number, little endian, "ARM\x64" */
+    u32 res5;			/* reserved (used for PE COFF offset) */
+    ====================================
+     */
+    const AARCH64_KERNEL_LOAD_ADDR: usize = 0x80000;
+    const AARCH64_MAGIC_NUMBER: u32 = 0x644d5241;
+    let mut kernel_load_offset = AARCH64_KERNEL_LOAD_ADDR;
+
+    /* Look for the magic number inside the elf header. */
+    let magic_offset_header = 2 * mem::size_of::<u32>() + 6 * mem::size_of::<u64>(); // This should total 56.
+
+    kernel_image
+        .seek(SeekFrom::Start(magic_offset_header as u64))
+        .map_err(|_| Error::SeekKernelImage)?;
+    let mut magic_number: u32 = 0;
+    unsafe {
+        sys_util::read_struct(kernel_image, &mut magic_number)
+            .map_err(|_| Error::ReadProgramHeader)?
+    }
+    if magic_number != AARCH64_MAGIC_NUMBER {
+        return Err(Error::InvalidElfMagicNumber);
+    }
+
+    /* Look for the `text_offset` from the elf header. */
+    let text_offset = 2 * mem::size_of::<u32>() as u64;
+    kernel_image
+        .seek(SeekFrom::Start(text_offset)) // This should total 8.
+        .map_err(|_| Error::SeekKernelImage)?;
+    let mut hdrvals: [u64; 2] = [0; 2];
+    unsafe {
+        // `read_struct` is safe when reading a POD struct. It can be used and dropped without issue.
+        sys_util::read_struct(kernel_image, &mut hdrvals).map_err(|_| Error::ReadProgramHeader)?;
+    }
+    /* Following the boot protocol mentioned above. */
+    if hdrvals[1] != 0 {
```

Comment:
As per the documentation the image_size is only zero for kernel versions < v3.17 which is something that we don't really support anyway. Should we return an error here instead?

Reply:
`Prior to v3.17, the endianness of text_offset was not specified.  In
  these cases image_size is zero and text_offset is 0x80000 in the
  endianness of the kernel.  Where image_size is non-zero image_size is
  little-endian and must be respected.  Where image_size is zero,
  text_offset can be assumed to be 0x80000.
`
I think you are referring to this section from which I understand the following:
- prior to v3.17 image_size is zero and text_offset is 0x80000
- after v3.17 two scenarios are still possible:
    - image_size is non-zero , text offset is not necessarily 0x80000
    - image size is zero, text offset is 0x80000
This is what my code reflects.
We have (or should have) other ways of checking kernel compatibility (i.e check_cap for kvm related compatibility).